File: LAB 3/MIDPOINT_ELLIPSE_ALGO.m
clc; clear; close all;

RX = 15;
RY = 8;
CX = 25;
CY = 15;

figure;
hold on;
grid on;
axis equal;
title('Midpoint Ellipse Algorithm');
xlabel('X-axis');
ylabel('Y-axis');

X = 0;
Y = RY;
RX2 = RX * RX;
RY2 = RY * RY;
TWO_RX2 = 2 * RX2;
TWO_RY2 = 2 * RY2;

PX = 0;
PY = TWO_RX2 * Y;

X_VALS = [];
Y_VALS = [];

% Region 1
P = round(RY2 - (RX2 * RY) + (0.25 * RX2));
while PX < PY
    X_VALS = [X_VALS CX+X CX-X CX+X CX-X];
    Y_VALS = [Y_VALS CY+Y CY+Y CY-Y CY-Y];
    X = X + 1;
    PX = PX + TWO_RY2;
    if P < 0
        P = P + RY2 + PX;
    else
        Y = Y - 1;
        PY = PY - TWO_RX2;
        P = P + RY2 + PX - PY;
    end
end

% Region 2
P = round(RY2 * (X + 0.5)^2 + RX2 * (Y - 1)^2 - RX2 * RY2);
while Y >= 0
    X_VALS = [X_VALS CX+X CX-X CX+X CX-X];
    Y_VALS = [Y_VALS CY+Y CY+Y CY-Y CY-Y];
    Y = Y - 1;
    PY = PY - TWO_RX2;
    if P > 0
        P = P + RX2 - PY;
    else
        X = X + 1;
        PX = PX + TWO_RY2;
        P = P + RX2 - PY + PX;
    end
end

plot(X_VALS, Y_VALS, 'ks', 'MarkerSize', 4, 'MarkerFaceColor', 'k');
plot(CX, CY, 'ro', 'MarkerSize', 8, 'MarkerFaceColor', 'r');

File: LAB 3/Python/MIDPOINT_ELLIPSE_ALGO.py
import matplotlib.pyplot as plt
import numpy as np

RX = 15
RY = 8
CX, CY = 25, 15

plt.figure()
plt.grid(True)
plt.axis('equal')
plt.title('Midpoint Ellipse Algorithm')
plt.xlabel('X-axis')
plt.ylabel('Y-axis')

X = 0
Y = RY
RX2 = RX * RX
RY2 = RY * RY
TWO_RX2 = 2 * RX2
TWO_RY2 = 2 * RY2

PX = 0
PY = TWO_RX2 * Y

X_VALS = []
Y_VALS = []

def plot_symmetric_points(X_VALS, Y_VALS, CX, CY, X, Y):
    points = [(X, Y), (-X, Y), (X, -Y), (-X, -Y)]
    for PX, PY in points:
        X_VALS.append(CX + PX)
        Y_VALS.append(CY + PY)
    return X_VALS, Y_VALS

P = round(RY2 - (RX2 * RY) + (0.25 * RX2))
while PX < PY:
    X_VALS, Y_VALS = plot_symmetric_points(X_VALS, Y_VALS, CX, CY, X, Y)
    X += 1
    PX += TWO_RY2
    if P < 0:
        P += RY2 + PX
    else:
        Y -= 1
        PY -= TWO_RX2
        P += RY2 + PX - PY

P = round(RY2 * (X + 0.5) ** 2 + RX2 * (Y - 1) ** 2 - RX2 * RY2)
while Y >= 0:
    X_VALS, Y_VALS = plot_symmetric_points(X_VALS, Y_VALS, CX, CY, X, Y)
    Y -= 1
    PY -= TWO_RX2
    if P > 0:
        P += RX2 - PY
    else:
        X += 1
        PX += TWO_RY2
        P += RX2 - PY + PX

plt.plot(X_VALS, Y_VALS, 'ks', markersize=4, markerfacecolor='k')
plt.plot(CX, CY, 'ro', markersize=8, markerfacecolor='r')
plt.show()
